doc/man/ejsmod.1
.TH EJSMOD "1" "March 2009" "ejsmod" "User Commands"
.SH NAME
ejsmod \- Ejscript Module Generator
.SH SYNOPSIS
.B ejsmod
[\fI--cslots\fR] 
[\fI--empty\fR]
[\fI--html path\fR] 
[\fI--listing\fR]
[\fI--searchPath ejsPath\fR]
[\fI--showBuiltin\fR]
[\fI--showDebug\fR]
[\fI--version\fR]
modules ...
.SH DESCRIPTION
The \fBejsmod\fR command is a utility program for Ejscript module files to generate symbolic assembler listings, 
slot defintion files for native classes and HTML documentation.
.SH LISTINGS
.PP
To generate symbolic assembler listings, compile your source using the \fBejsc\fR command with the --debug switch. Then
invoke \fBejsmod\fR with the \fB--listing\fR switch and supply the required modules. An assembler listing 
file with a ".lst" extension will be generated for each corresponding logical module directive. 
.PP
The listing files interleave disassembled byte code with the relevant source
code statements. The listings describe each module and its constituent functions and initializers. The 
bindings for all types and the constant pool are also listed.
.SH DOCUMENTATION
To generate HTML documentation for your code, first compile using the \fBejsc\fR command with the --doc switch. Then
invoke \fBejsmod\fR and use the \fB--html\fR switch with the directory where you want the documentation generated.
.PP
The HTML documentation will include all modules, namespaces, types, functions, properties and variable declarations.
.SH BINDING SLOT FILES
Ejscript supports the creation of native types which are coded in the C language. Native types are useful for 
core system types and for classes that must run at near binary speed. To facilitate creating native types, \fBejsmod\fR
generates slot binding definition headers which define symbolic offsets for declarations. See the Native class
samples for further information.
.SH OPTIONS
.TP
\fB\--cslots\fR
Generate slot binding definition headers for the input modules. 
.TP
\fB\--empty\fR
Start with an empty interpreter without the core language types such as Object, Array and Number. This option is 
used to build the foundation ejs.mod module which contains the core system types.
.TP
\fB\--html dir\fR
Generate HTML documentation to the specified directory. The source files must have been compiled by \fBejsc\fR 
with the --doc switch to include the documentation strings in the module files.
.TP
\fB\--listing\fR
Generate symbolic assembler files for each input module.
.TP
\fB\--searchPath ejsPath\fR
Override the module search path. The module search path is a set of directories that the \fBejsmod\fR command will use
when locating and loading Ejscript modules. Given a module named "a.b.c" in a script, \fBejsmod\fR will use the following 
search strategy to locate the module:
.IP
1. Search for a module file named "a.b.c.mod"
.IP
2. Search for a module file named "a/b/c.mod"
.IP
3. Search for a module file named "a.b.c.mod" in EJSPATH
.IP
4. Search for a module file named c.mod in EJSPATH
.IP
The search path is initially specified via the environment variable EJSPATH and may be overridden via the --searchPath ejsPath
switch. EJSPATH and the ejsPath command line value are similar to the system PATH formats. On windows, path segments 
are separated by ";" and on Linux, Unix, FreeBSD and MAC, the path segments are separated by ":" delimiters.
.TP
\fB\--showDebug\fR
Show debug instructions in the symbolic assembler listings.
.TP
\fB\--showBuiltin\fR
Show built-in properties in HTML documentation, assembler listings and slot binding files.
.TP
\fB\--version\fR
Print the \fBejsmod\fR command version and exit.
.SH BACKGROUND
Ejscript is an enhanced implementation of the JavaScript language for use in embedded applications such as web 
servers, embedded devices and mobile phones. It is especially well suited for Server-Side JavaScript web applications.
.PP
Ejscript is a dynamic, interpreted, object-oriented scripting language that supports classes, objects, exceptions, 
statements, expressions and a powerful suite of data types.
.SH "REPORTING BUGS"
Report bugs to dev@embedthis.com.
.SH COPYRIGHT
Copyright \(co 2004-2009 Embedthis Software.  Ejscript is a trademark of Embedthis Software.
.br
.SH "SEE ALSO"
ejsc, ejsgen, ejs, ejsvm


src/ejsmod/ejsmod.c
/**
 *  ejsmod.c - Module manager 
 *
 *  Copyright (c) All Rights Reserved. See details at the end of the file.
 */

/********************************** Includes **********************************/

#include    "ejsMod.h"

/****************************** Forward Declarations **************************/

static int  process(EjsMod *mp, int argc, char **argv);
static void logger(MprCtx ctx, int flags, int level, const char *msg);
static int  setLogging(Mpr *mpr, char *logSpec);

/************************************ Code ************************************/

int main(int argc, char **argv)
{
    Mpr             *mpr;
    EjsMod          *mp;
    EjsService      *vmService;
    Ejs             *ejs;
    char            *argp, *searchPath;
    int             nextArg, err, flags;

    err = 0;
    searchPath = 0;
    
    /*
     *  Create the Embedthis Portable Runtime (MPR) and setup a memory failure handler
     */
    mpr = mprCreate(argc, argv, ejsMemoryFailure);
    mprSetAppName(mpr, mprGetBaseName(argv[0]), 0, 0);

    /*
     *  Allocate the primary control structure
     */
    mp = mprAllocObjZeroed(mpr, EjsMod);
    if (mp == 0) {
        return MPR_ERR_NO_MEMORY;
    }
    mp->lstRecords = mprCreateList(mp);
    mp->blocks = mprCreateList(mp);
    mp->docDir = ".";
    
    for (nextArg = 1; nextArg < argc; nextArg++) {
        argp = argv[nextArg];
        if (*argp != '-') {
            break;
        }
        if (strcmp(argp, "--cslots") == 0) {
            mp->cslots = 1;
            mp->genSlots = 1;
            
        } else if (strcmp(argp, "--empty") == 0) {
            mp->empty = 1;
            
        } else if (strcmp(argp, "--error") == 0) {
            /*
             *  Undocumented switch
             */
            mp->exitOnError++;
            mp->warnOnError++;
            
        } else if (strcmp(argp, "--jslots") == 0) {
            /*
             *  This command is currently not documented until the JVM is released.
             */
            mp->jslots = 1;
            mp->genSlots = 1;

        } else if (strcmp(argp, "--html") == 0) {
#if BLD_FEATURE_EJS_DOC
            if (nextArg >= argc) {
                err++;
            } else {
                mp->docDir = argv[++nextArg];
                mp->html = 1;
            }
#else
            mprErrorPrintf(mpr, "Doc generation is not enabled. Reconfigure with --enable-doc");
            err++;
#endif
            
        } else if (strcmp(argp, "--listing") == 0) {
            mp->listing = 1;
            
        } else if (strcmp(argp, "--log") == 0) {
            /*
             *  Undocumented switch
             */
            if (nextArg >= argc) {
                err++;
            } else {
                setLogging(mpr, argv[++nextArg]);
            }

        } else if (strcmp(argp, "--searchpath") == 0) {
            if (nextArg >= argc) {
                err++;
            } else {
                searchPath = argv[++nextArg];
            }

        } else if (strcmp(argp, "--showDebug") == 0) {
            mp->showDebug++;

        } else if (strcmp(argp, "--showBuiltin") == 0) {
            mp->showBuiltin = 1;
            
        } else if (strcmp(argp, "--version") == 0 || strcmp(argp, "-V") == 0) {
            mprErrorPrintf(mpr, "%s %s"
                "Copyright (C) Embedthis Software 2003-2009\n"
                "Copyright (C) Michael O'Brien 2003-2009\n",
               BLD_NAME, BLD_VERSION);  
            exit(0);

        } else if (strcmp(argp, "--warn") == 0) {
            /*
             *  Undocumented switch
             */
            mp->warnOnError++;

        } else if (strcmp(argp, "--xml") == 0) {
#if BLD_FEATURE_EJS_DOC
            mp->xml = 1;
#else
            mprErrorPrintf(mpr, "Doc generation is not enabled. Reconfigure with --enable-doc");
            err++;
#endif
        
        } else {
            err++;
            break;
        }
    }
    
    if (argc == nextArg) {
        err++;
    }
    
    if (mp->genSlots == 0 && mp->listing == 0 && mp->html == 0 && mp->xml == 0) {
        mp->listing = 1;
    }

    if (err) {
        /*
         *  Examples:
         *      ejsmod file.mod                              # Defaults to --listing
         *      ejsmod --out listing.lst embedthis.mod 
         */
        mprErrorPrintf(mpr, 
            "Usage: %s [options] modules ...\n"
            "  Ejscript module manager options:\n"
            "  --cslots              # Generate a C slot definitions file\n"
            "  --empty               # Create empty interpreter\n"
            "  --html dir            # Generate HTML documentation to the specified directory\n"
            "  --listing             # Create assembler listing files (default)\n"
            "  --searchpath ejsPath  # Module file search path\n"
            "  --showBuiltin         # Show builtin properties\n"
            "  --showDebug           # Show debug instructions\n"
            "  --version             # Emit the program version information\n\n", mpr->name);
        return -1;
    }

    /*
     *  Need an interpreter to load modules
     */
    vmService = ejsCreateService(mpr); 
    if (vmService == 0) {
        return MPR_ERR_NO_MEMORY;
    }

    if (searchPath) {
        ejsSetSearchPath(vmService, searchPath);
    }
    
    flags = EJS_FLAG_COMPILER | EJS_FLAG_NO_EXE;
    if (mp->empty) {
        flags |= EJS_FLAG_EMPTY;
    }
    if (mp->html || mp->xml) {
        flags |= EJS_FLAG_DOC;
    }
    ejs = ejsCreate(vmService, NULL, flags);
    if (ejs == 0) {
        return MPR_ERR_NO_MEMORY;
    }
    mp->ejs = ejs;

    if (nextArg < argc) {

        /*
         *  Process the module files supplied on the command line.
         */
        if (process(mp, argc - nextArg, &argv[nextArg]) < 0) {
            err++;
        }
    }

    if (mp->errorCount > 0) {
        err = -1;
    }
    mprFree(mpr);
    return err;
}



static int process(EjsMod *mp, int argc, char **argv)
{
    MprList     *modules, *allModules;
    EjsModule   *module;
    void        (*callback)(struct Ejs *ejs, int kind, ...);
    int         i, next;

    callback = (mp->listing) ? emListingLoadCallback : 0;
    allModules = mprCreateList(mp);
    
    /*
     *  For each module on the command line
     */
    for (i = 0; i < argc && !mp->fatalError; i++) {
        mp->ejs->userData = mp;
        module = 0;
        if (!mprAccess(mp, argv[i], R_OK)) {
            mprError(mp, "Can't access module %s", argv[i]);
            return EJS_ERR;
        }
        if ((modules = ejsLoadModule(mp->ejs, argv[i], 0, callback, EJS_MODULE_DONT_INIT)) == 0) {
            mprError(mp, "Can't load module %s\n%s", argv[i], ejsGetErrorMsg(mp->ejs, 0));
            return EJS_ERR;
        }
        
        if (mp->genSlots) {
            /*
             *  Create the slot files for all modules loaded
             */
            for (next = 0; (module = mprGetNextItem(modules, &next)) != 0; ) {
                emCreateSlotFiles(mp, module);
            }
        }
        mprAppendList(allModules, modules);
        mprFree(modules);
    }
        
#if BLD_FEATURE_EJS_DOC
    if (mp->html || mp->xml) {
        emCreateDoc(mp);
    }
#endif

    return 0;
}



static int setLogging(Mpr *mpr, char *logSpec)
{
    MprFile     *file;
    char        *levelSpec;
    int         level;

    level = 0;

    if ((levelSpec = strchr(logSpec, ':')) != 0) {
        *levelSpec++ = '\0';
        level = atoi(levelSpec);
    }

    if (strcmp(logSpec, "stdout") == 0) {
        file = mpr->fileService->console;
    } else {
        if ((file = mprOpen(mpr, logSpec, O_WRONLY, 0664)) == 0) {
            mprErrorPrintf(mpr, "Can't open log file %s\n", logSpec);
            return EJS_ERR;
        }
    }

    mprSetLogLevel(mpr, level);
    mprSetLogHandler(mpr, logger, (void*) file);

    return 0;
}



static void logger(MprCtx ctx, int flags, int level, const char *msg)
{
    Mpr         *mpr;
    MprFile     *file;
    char        *prefix;

    mpr = mprGetMpr(ctx);
    file = (MprFile*) mpr->logHandlerData;
    prefix = mpr->name;

    while (*msg == '\n') {
        mprFprintf(file, "\n");
        msg++;
    }

    if (flags & MPR_LOG_SRC) {
        mprFprintf(file, "%s: %d: %s\n", prefix, level, msg);

    } else if (flags & MPR_ERROR_SRC) {
        /*
         *  Use static printing to avoid malloc when the messages are small.
         *  This is important for memory allocation errors.
         */
        if (strlen(msg) < (MPR_MAX_STRING - 32)) {
            mprStaticPrintf(file, "%s: Error: %s\n", prefix, msg);
        } else {
            mprFprintf(file, "%s: Error: %s\n", prefix, msg);
        }

    } else if (flags & MPR_FATAL_SRC) {
        mprFprintf(file, "%s: Fatal: %s\n", prefix, msg);
        
    } else if (flags & MPR_RAW) {
        mprFprintf(file, "%s", msg);
    }
    
    if (flags & (MPR_ERROR_SRC | MPR_FATAL_SRC | MPR_ASSERT_SRC)) {
        mprBreakpoint();
    }
}


/*
 *  @copy   default
 *  
 *  Copyright (c) Embedthis Software LLC, 2003-2009. All Rights Reserved.
 *  Copyright (c) Michael O'Brien, 1993-2009. All Rights Reserved.
 *  
 *  This software is distributed under commercial and open source licenses.
 *  You may use the GPL open source license described below or you may acquire 
 *  a commercial license from Embedthis Software. You agree to be fully bound 
 *  by the terms of either license. Consult the LICENSE.TXT distributed with 
 *  this software for full details.
 *  
 *  This software is open source; you can redistribute it and/or modify it 
 *  under the terms of the GNU General Public License as published by the 
 *  Free Software Foundation; either version 2 of the License, or (at your 
 *  option) any later version. See the GNU General Public License for more 
 *  details at: http://www.embedthis.com/downloads/gplLicense.html
 *  
 *  This program is distributed WITHOUT ANY WARRANTY; without even the 
 *  implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
 *  
 *  This GPL license does NOT permit incorporating this software into 
 *  proprietary programs. If you are unable to comply with the GPL, you must
 *  acquire a commercial license to use this software. Commercial licenses 
 *  for this software and support services are available from Embedthis 
 *  Software at http://www.embedthis.com 
 *  
 *  @end
 */
